src/App/Fossa/YamlDeps.hs

```diff
@@ -0,0 +1,196 @@
+
+    addEmptyDep :: Locator -> (Locator, SourceUnitDependency)
+    addEmptyDep loc = (loc, SourceUnitDependency loc [])
+
+toAdditionalData :: NE.NonEmpty CustomDependency -> AdditionalDepData
+toAdditionalData deps = AdditionalDepData {userDefinedDeps = map tosrc $ NE.toList deps}
+  where
+    tosrc CustomDependency {..} =
+      SourceUserDefDep
+        { srcUserDepName = customPackage,
+          srcUserDepVersion = customVersion,
+          srcUserDepLicense = customLicense,
+          srcUserDepDescription = customDescription,
+          srcUserDepUrl = customUrl
+        }
+
+hasNoDeps :: YamlDependencies -> Bool
+hasNoDeps YamlDependencies{..} = all (0 ==)
+  [ length referencedDependencies,
+    length customDependencies
+  ]
+
+data YamlDependencies = YamlDependencies
+  { referencedDependencies :: [ReferencedDependency],
+    customDependencies :: [CustomDependency]
+  } deriving (Eq, Ord, Show)
+
+data ReferencedDependency = ReferencedDependency
+  { locDepPackage :: Text,
+    locDepType :: DepType,
+    locDepVersion :: Maybe Text
+  }
+  deriving (Eq, Ord, Show)
+
+data CustomDependency = CustomDependency
+  { customPackage :: Text,
+    customVersion :: Text,
+    customLicense :: Text,
+    customDescription :: Maybe Text,
+    customUrl :: Maybe Text
+  } deriving (Eq, Ord, Show)
+
+instance FromJSON YamlDependencies where
+  parseJSON = withObject "YamlDependencies" $ \obj ->
+    YamlDependencies <$ (obj .:? "version" >>= isMissingOr1)
+      <*> (obj .:? "referenced-dependencies" .!= [])
+      <*> (obj .:? "custom-dependencies" .!= [])
+    where
+      isMissingOr1 :: Maybe Int -> Parser ()
+      isMissingOr1 (Just x) | x /= 1 = fail $ "Invalid fossa-deps version: " <> show x
+      isMissingOr1 _ = pure ()
+
+depTypeParser :: Text -> Parser DepType
+depTypeParser text = case depTypeFromText text of
+  Just t -> pure t
+  Nothing -> fail $ "dep type: " <> unpack text <> " not supported"
+
+instance FromJSON ReferencedDependency where
+  parseJSON = withObject "ReferencedDependency" $ \obj ->
+    ReferencedDependency <$> obj .: "package"
+      <*> (obj .: "type" >>= depTypeParser)
+      <*> obj .:? "version"
+      <* forbidMembers "referenced dependencies" ["license", "description", "url"] obj
+
+instance FromJSON CustomDependency where
+  parseJSON = withObject "CustomDependency" $ \obj ->
+    CustomDependency <$> obj .: "package"
+      <*> obj .: "version"
+      <*> obj .: "license"
+      <*> obj .:? "description"
+      <*> obj .:? "url"
+      <* forbidMembers "custom dependencies" ["type"] obj
+
+-- | Parser insert to prevent specific fields from being used in parsers
+-- Primarily useful for rejecting aeson fields which should be reported with custom error messages
+--
+-- >  parseJSON = withObject "MyDataType" $ \obj ->
+-- >   MyDataType <$> obj .: "my-data-field"
+-- >     <* forbidMembers "Custom error message" ["badfield1", "badfield2"] obj
+forbidMembers :: Text -> [Text] -> Object -> Parser ()
```

Comment:
This is awesome, eases most of the concerns we had about users entering incorrect data and expecting it to behave differently.

Reply:
+1 -- this is great

Reply:
Should this be promoted to `Data.Aeson.Extra`?  Do we ever expect to use it outside of `fossa-deps`?  We can always move it later.

Also worth noting that generic aeson parsing does this automatically (if given a certain option), and we may need to try using that along with `-XDuplicateRecordFields` and `-Werror -Wambiguous-fields`.  Since we haven't taken direct input from files yet, we haven't experimented with generic JSON implementations.

Reply:
I think it's useful enough to move to `Data.Aeson.Extra`

re: generic aeson: I'd like to avoid using it if possible. I like the explicit nature of `FromJSON` instances -- and it's usually minimal overhead to write those instances. It doesn't help that `DuplicateRecordFields` behaves really poorly in most places

It might be worthwhile to add the opposite of `forbidMembers` -- `expectOnlyMembers` or whatever -- that would fail if there were any other fields in the object

Reply:
Though it would be easy to implement.